src/app/machines/views/MachineList/MachineListHeader/MachineListHeader.tsx

```diff
@@ -28,24 +29,56 @@ import { getNodeActionTitle } from "app/store/utils";
 
 type Props = {
   headerContent: MachineHeaderContent | null;
+  searchFilter: string;
   setSearchFilter: SetSearchFilter;
   setHeaderContent: MachineSetHeaderContent;
 };
 
 export const MachineListHeader = ({
   headerContent,
+  searchFilter,
   setSearchFilter,
   setHeaderContent,
 }: Props): JSX.Element => {
   const location = useLocation();
-  const selectedMachines = useSelector(machineSelectors.selected);
+  const selected = useSelector(machineSelectors.selected);
+  let selectedState = useSelector(machineSelectors.selectedMachines);
   const hasSelection = useHasSelection();
   const [tagsSeen, setTagsSeen] = useStorageState(
     localStorage,
     "machineViewTagsSeen",
     false
   );
-  const { machineCount, machineCountLoading } = useFetchMachineCount();
+  // Get the count of all machines that match the current filters.
+  const { machineCount, machineCountLoading } = useFetchMachineCount(
+    FilterMachineItems.parseFetchFilters(searchFilter)
+  );
+  let selectedCount = 0;
+  // Shallow clone the selected state so that object can be modified.
+  let selectedMachines = selectedState ? { ...selectedState } : null;
+  // Remove selected items from the filters to send to the API. We can count
+  // them client side and filters are combined with AND which we don't want to do when
+  // there are selected groups and items (otherwise it will be counting the
+  // machines that match both the groups and the items).
+  if (selectedMachines && "items" in selectedMachines) {
+    selectedCount += selectedMachines.items?.length ?? 0;
+    delete selectedMachines.items;
+  }
+  // Get the count of machines in selected groups or filters.
+  const filters = selectedToFilters(selectedMachines);
+  const {
+    machineCount: fetchedSelectedCount,
+    machineCountLoading: selectedLoading,
+  } = useFetchMachineCount(filters);
+  // Only add the count if there are filters as sending `null` filters
+  // to the count API will return a count of all machines.
+  if (filters) {
+    selectedCount += fetchedSelectedCount;
+  }
+  const onlyHasItems =
+    !!selectedMachines &&
+    selectedCount > 0 &&
+    (!("groups" in selectedMachines) || !selectedMachines?.groups?.length);
```

Comment:
Do we have a use case for getting a count that matches both the groups and the items?

It seems like this should either already be a part of the `useFetchMachineCount` hook or abstracted to another one, e.g. `useSelectedMachineCount`, along with logic below - `machineCountLoading || onlyHasItems ? false : selectedLoading`).